### ISSUE TITLE
azure tests still making http calls?

### ISSUE THREAD

**Issue by kapilt** [capitalone/cloud-custodian#2338]
we've been debugging slow test execution in our environment (public env tests do about ~50s or less) in our env its about 1200s, the tests seem to execute speedily till we get to azure stuff then, they start going really, really slow. if we control-c to break out, we always see https calls to azure services in the traceback

@stefangordon ^ when its replay mode, should these tests be making any network calls?

**Comment by stefangordon** [comment]
It doesn't actually block HTTP calls, and it is possible it is attempting token auth.  There is probably some cleanup to do here I will look at.

Can you clarify exactly how you are running them?  Are you just running tox or some other method?  (Mostly want to verify what AZURE_* environment variables you have).

**Comment by kapilt** [comment]
```.....^CERROR: KEYBOARDINTERRUPT                                                                                                                            DEBUG:msrest.pipeline:Configuring request: timeout=100, verify=True, cert=None                                                                                         DEBUG:msrest.pipeline:Configuring proxies: ''                                                                                          
DEBUG:msrest.pipeline:Evaluate proxies against ENV settings: True   
                
DEBUG:msrest.pipeline:Configuring redirects: allow=True, max=30 
                                                                                                       
DEBUG:msrest.pipeline:Configuring retry: max_retries=4, backoff_factor=0.8, max_backoff=90    
   DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): management.azure.com 
                                                                                  
DEBUG:urllib3.util.retry:Incremented Retry for (url='/subscriptions/aa98974b-omit/operationresults/eyJqb2JJZCI6IlJFU09VUkNFR1JPVVBERUxFVElPTkpPQ
i1URVNUOjVGRU1QVFlSRy1TT1VUSENFTlRSQUxVUyIsImpvYkxvY2F0aW9uIjoic291dGhjZW50cmFsdXMifQ?api-version=2017-05-10'): Retry(total=3, connect=3, read=4, redirect=None, status=None)   
                 
WARNING:urllib3.connectionpool:Retrying (Retry(total=3, connect=3, read=4, redirect=None, 
status=None)) after connection broken by 'NewConnectionError('<urllib3.connec
tion.VerifiedHTTPSConnection object at 0x10d859610>: Failed to establish a new connection: [Errno 60] Operation timed out',)': /subscriptions/aa98974b-omit/operationresults/eyJqb2JJZCI6IlJFU09VUkNFR1JPVVBERUxFVElPTkpPQi1URVNUOjVGRU1QVFlSRy1TT1VUSENFTlRSQUxVUyIsImpvYkxvY2F0aW9uIjoic291dGhjZW50cmFsdXMifQ?api-version=2017-05-10

DEBUG:urllib3.connectionpool:Starting new HTTPS connection (2): management.azure.com       
                                                                            
DEBUG:urllib3.util.retry:Incremented Retry for (url='/subscriptions/aa98974--omit/operationresults/eyJqb2JJZCI6IlJFU09VUkNFR1JPVVBERUxFVElPTkpPQi1URVNUOjVGRU1QVFlSRy1TT1VUSENFTlRSQUxVUyIsImpvYkxvY2F0aW9uIjoic291dGhjZW50cmFsdXMifQ?api-version=2017-05-10'): Retry(total=2, connect=2, read=4, redirect=None, status=None)         

WARNING:urllib3.connectionpool:Retrying (Retry(total=2, connect=2, read=4, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connec
tion.VerifiedHTTPSConnection object at 0x10dac5050>: Failed to establish a new connection: [Errno 60] Operation timed out',)': /subscriptions/aa98974b-5d2a-DEBUG:msrest.pipeline:Configuring request: timeout=100, verify=True, cert=None     
```

**Comment by kapilt** [comment]
just running via tox, ideally we wouldn't need to be making http calls from the test suite when replaying.

**Comment by kapilt** [comment]
the alternative is we separate out the azure env from the default tox setup to separate tox targets.

**Comment by stefangordon** [comment]
I did confirm that VCR will just raise an exception on attempted HTTP requests if there is already a cassette associated to the test, so it shouldn't be possible for the tests to reach out to the internet (unless the cassette was not present, but then the test would fail).     Because the subscription ID in your trace isn't the subscription ID used in the unit tests, I'm suspicious it's coming from something else??

```
    AZURE_ACCESS_TOKEN=fake_token
    AZURE_SUBSCRIPTION_ID=ea42f556-5106-4743-99b0-c129bfa71a47
```

**Comment by kapilt** [comment]
i'm not entirely sure why we're seeing such slow downs in our corporate env. on a personal machine, i was able to verify the azure tests aren't doing outbound net traffic, ie. they work with networking on my machine turned off. i'll continue to investigate but afaics the azure tests are not at issue.

**Comment by stefangordon** [comment]
We don't pin the Azure SDK versions and I have seen little things that make me suspicious different versions handle the VCR HTTP hook differently but I haven't quite reproduced exactly yet.  It is worth verifying that `pip show azure` shows 3.0.0

I'm also going to slightly tweak the way we read and set env variables so that the test run can clear `AZURE_CLI_AUTH` vars that would override its test tokens.

**Comment by kapilt** [comment]
doh.. i think it was that poc test module at the top level of the azure pkg

**Comment by stefangordon** [comment]
Ah I was just thinking about that - I'll delete it next time I send a PR.

**Comment by stefangordon** [comment]
#2342

**Comment by kapilt** [comment]
thanks, i think i should apologize for the noise and confusion here, i'm convinced our issues are not related to azure pkg at all. i'm able to reproduce the slow downs in our corp env even with all the tools subdir tests disabled. i'll continue to investigate, but closing this issue out. /me reaches for git bisect.

**Comment by kapilt** [comment]
fwiw, git bisect pointed to https://github.com/capitalone/cloud-custodian/commit/4e9f802dd7d46842ecf6e2eb40b195b90612a28f  as when the slow down originated, but the why is proving a bit hard to analyze further, things like running tests single process make the problem go away, as do adding a profiler into the mix when running tests in parallel. My theory at the moment is its related to the cli tests and account id acquisition from the cloud provider, which that revision missed, those cli tests capture stdout/stderr which is why they didn't pop previously. Why this behavior is highly differential to our corporate env is another oddity. I'll continue to investigate.

**Comment by kapilt** [comment]
confirmed theory,  and posted fix in #2343
